### DISCUSSION TITLE
Power#State rules aren't triggered on boot

### DISCUSSION THREAD

**Discussion by shantur** [arendst/Tasmota#17979] [Ideas]
### PROBLEM DESCRIPTION
Power#State rules aren't triggered on Boot.

I have an ESP8265/66 4 Relay device running Tasmota 12.4.0 and my requirement is 

- Whenever Power1 is turned On, turn on Power2
- Power2 is configured to be a Pulse so it turns off after a few seconds.

I can create a rule for this such as 

```
rule1 on Power1#state=1 
  do power2 on endon
```

This rule works fine for normal use but when the device is powered off and powered on the Power1 is restored with PowerOnState but the `Power1#State` isn't triggered on boot which leads to unexpected state.

To solve this another idea could be to use 

```
rule1 on Power1#Boot=1
  do power2 on endon
on Power1#State=1
  do power2 on endon
```
This solves the problem on the boot, but the Power1#Boot=1 triggers on the Warm restarts of the device leading to again unexpected Power2 to be turned on when Power1 was not turned On.

As I understand, the logic behind Power1#State rule is to trigger it on a state change and on a cold boot the power is restored and hence not a state change. But from a power perspective, when the device powers down Power1 is turned off by loss of power and when powered up again Power1 is turned On leading to a state change.

Possible solution is to trigger Power1#State rule on Power restore.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): eWeLink 4CH
- [x] Tasmota binary firmware version number used: 12.4.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: Tasmotizer
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
01:51:43.563 RSL: RESULT = {"NAME":"eWeLink 4CH","GPIO":[160,0,0,0,226,225,161,162,224,288,163,227,0,0],"FLAG":0,"BASE":18}
01:51:43.812 RSL: RESULT = {"Module":{"0":"eWeLink 4CH"}}
01:51:44.018 RSL: RESULT = {"GPIO0":{"160":"Switch1"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"226":"Relay3"},"GPIO5":{"225":"Relay2"},"GPIO9":{"161":"Switch2"},"GPIO10":{"162":"Switch3"},"GPIO12":{"224":"Relay1"},"GPIO13":{"288":"Led1"},"GPIO14":{"163":"Switch4"},"GPIO15":{"227":"Relay4"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:

01:52:33.431 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"MaasPowerControl1","FriendlyName":["Tasmota","","",""],"Topic":"tasmota_49A15A","ButtonTopic":"0","Power":3,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
01:52:33.474 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Power On","Uptime":"0T01:11:05","StartupUTC":"2023-02-17T23:41:28","Sleep":50,"CfgHolder":4617,"BootCount":24,"BCResetTime":"2023-02-07T15:59:59","SaveCount":177,"SaveAddress":"F5000"}}
01:52:33.511 RSL: STATUS2 = {"StatusFWR":{"Version":"12.4.0(tasmota)","BuildDateTime":"2023-02-16T12:12:54","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8285N08","CR":"353/699"}}
01:52:33.549 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["RRI",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
01:52:33.591 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":632,"Free":368,"Heap":24,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"14325E","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
01:52:33.630 RSL: STATUS5 = {"StatusNET":{"Hostname":"MaasPowerControl1","IPAddress":"192.168.6.46","Gateway":"192.168.6.1","Subnetmask":"255.255.254.0","DNSServer1":"192.168.6.1","DNSServer2":"0.0.0.0","Mac":"48:55:19:49:A1:5A","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
01:52:33.670 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_49A15A","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
01:52:33.710 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-02-18T00:52:33","Local":"2023-02-18T01:52:33","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"07:53","Sunset":"18:14"}}
01:52:33.748 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-02-18T01:52:33","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON"}}
01:52:33.784 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-02-18T01:52:33","Uptime":"0T01:11:05","UptimeSec":4265,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER1":"ON","POWER2":"ON","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"RRI","BSSId":"7E:83:C2:2D:5F:D9","Channel":11,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":3,"Downtime":"0T00:00:31"}}}
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
On a dual relay device

1. create a rule for this such as 

```
rule1 on Power1#state=1 
  do power2 on endon
```
2. Enable PulseTime for Relay2 by `PulseTime2 20`
3. Power1 On
4. Power re-cycle the device

### EXPECTED BEHAVIOUR
When Power1 is restored, Power2 should be pulsed.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
The suggestion of changing current behavior to no longer be when Tasmota works to change the power power state would be a breaking change.

**Comment by gitmeister** [comment]
At Power-On aren't the State variables in an unknown state and need to be evaluated so that rules (ex. On Switch1#State=) can trigger correctly?

**Reply by sfromis** [reply]
`xxxx#State` is not a state "variable", but an event about something changing, as it happens. At boot time, you do also have triggers like `Power1#Boot` or `Switch1#Boot` to get the state (as seen by Tasmota) at boot, but this is of course different triggers from the regular state-*change* triggers.

**Comment by gitmeister** [comment]
I guess my point is that the state has changed at power-on since you'd expect an event to be triggered because the xxxx#State went from Unknown->On or Off.

EDIT: Looking at the code in "support_switch_v4.ino", it seems the option would be in SwitchInit():
`244                Switch.last_state[i] = digitalRead(Pin(GPIO_SWT1, i)); `

If an "Unknown" code (say, 255) could be substituted there, then SwitchHandler() would see a change in the test:
`button != Switch.last_state[i]`

and invoke a change-of-state event upon power-on and prevent rules from "hanging" waiting for an state-change.

**Reply by sfromis** [reply]
Changed means from on to off, or vice versa. That's how Tasmota works.
You could argue that restoring previous state to same could count as a state change too, but that would be a breaking change.

Of course, "restoring state" is about Powerx, not about Switchx.

**Comment by gitmeister** [comment]
Maybe our signals are crossed but I added some code above that addresses my issue. I'll add an obscure Option to trigger it.

**Reply by sfromis** [reply]
Simpler to just have a `xxxxx#Boot` trigger to pick up boot state.
